### PR TITLE
Dropdown: Screen reader now reads out required for required dropdowns

### LINKED ISSUE #5162: Dropdown: Field 'Required' is not read by screen reader
Bug Report:
Package version(s): (5.66.0)
Browser and OS versions: Edge, Screen Reader(Narrator)
Priorities and help requested (not applicable if asking question):
Are you willing to submit a PR to fix? (No)
Requested priority: ( Normal)
Products/sites affected: (if applicable)

Describe the issue:
For implementation of 'Dropdown' using 'required' prop with 'label' prop, dropdown label is showing the (*) but narrator does not read Dropdown as 'Required' when dropdown receives focus. Similar case with Date Picker.

The behavior should be similar like TextField.

Actual behavior:
Screen reader is not reading dropdown as required (if 'required' prop is specified) when it receives focus.

Expected behavior:
If 'required' prop is set for dropdown then Screen Reader should read Dropdown as 'Required' when focus reaches to this(Similar to TextField)
If applicable, please provide a codepen repro:

### PR DESCRIPTION
**Pull request by natalieethell**
#### Pull request checklist

- [X] Addresses an existing issue: Fixes #5162
- [X] Include a change request file using `$ npm run change`

#### Description of changes

Narrator wasn't reading out "required" for required dropdowns. There was a long discussion figuring out why "aria-required" was not being read out for the listbox element, and we found out it was because the focus was jumping to the list item inside it (which is not keyboard-focusable, but comes to Narrator as having keyboard focus). So, this fix involves only maintaining aria-activedescendant when the listbox is open, and unsetting the attribute when closed. This gives us the desired behavior.

We do lose just part of what used to be read out, but reading out "required" seems more important than reading out the selected item's index when the dropdown is closed.

**Edge + Narrator**

**Before:**
Basic dropdown example, Option a, <say-as interpret-as="spell">1</say-as> of 10, selected

**After:**
Basic dropdown example, Option a selected, requires selection contains 10 items,Required,
Option a, collapsed,

**Edge + NVDA
Chrome + NVDA**

**Before:**
Basic dropdown example list collapsed required Option a
Option a, 1 of 10

**After:**
Basic dropdown example list collapsed required Option a

#### Focus areas to test

Turn on Narrator and the Dropdown demo page in Edge. Navigate to the Required Dropdown example using the keyboard and ensure that the following is read out:

Basic dropdown example, Option a selected, requires selection contains 10 items,Required,
Option a, collapsed,

### CODE REVIEW

**Comment by cliffkoh** [comment]
The deploy link does not load for me.. :( @kenotron I think it is the same issue as last time...

**Comment by natalieethell** [comment]
@cliffkoh it's working now!

**Comment by cliffkoh** [comment]
For the non-required dropdowns, I hear "Option a selected. Selection contains 10 items". Doesn't "Selection contains 10 items" sound wrong to you? This almost feels like a regression (in the non-required dropdown case) from what it was formerly.

Having said that, I'm not too sure where the "10 items" come from since it is not on the `<div id={id}>`

For the required dropdown case, it seems to be fine because, if I add in my own punctuation, it can be interpreted as: "Basic dropdown example, Option a selected, requires selection. Contains 10 items, Required"

The inclusion of "requires" there completely changes how I would understand "selection contains 10 items"...

**Comment by cliffkoh** [comment]
Funny thing is, the NVDA readout (where it no longer reads 1 of 10 anymore - which almost feels like a regression in that sense) on the other hand, is more intuitive (because there's no more aria-activedescendent, there's no information on aria-size).

Note that aria-activedesendent that was there in the very first place to indicate that the 1 rendered option on screen is part of a set of options that you can change via up/down arrow keys. Hence, I actually feel the NVDA change in readout behavior is definitely a regression as well from this POV, in addition to the confusing Narrator experience with non-required Dropdowns...

**Comment by natalieethell** [comment]
I think the "10 items" comes from the `aria-describedby`, which points to the element that has `aria-setsize` set to 10.

If you think this feels like a regression, what we can do is only remove `aria-activedescendant` from the required dropdowns in their collapsed state. The non-required dropdowns can remain as is.

@cschlechty when you have a chance, could you check out this behavior change? I'm curious to hear what you think.

**Comment by cliffkoh** [comment]
What I am specifically unsure about is what is considered "standards compliant". I lean towards our existing implementation as being slightly more standards compliant (and NVDA's experience is better in that sense). What you proposed crossed my mind too :) It seems to result in best of both worlds with the screen readers (at least these two) today but the fact that we're doing this special handling hack is a telltale sign that we're not coding for standards compliance (at least one of the two must be non-standards compliant) :)

I wonder though if you can share our findings with the Narrator team (including the behavior in NVDA) and see if they can provide any feedback - that feedback given our findings/discussion here can help inform us on the way forward here.

Thanks Natalie, and thanks for the sustained effort on this! Greatly appreciated!

**Comment by cschlechty** [comment]
@natalieethell @cliffkoh it looks like dropdown mixes up some ARIA. I don't see why we're giving the launching control an inner "option" and setting posinset and setsize. Those should only be used in the actual options list, and really only needed if you lazy load your items. 

I agree with the activedescendant change.

**Comment by cliffkoh** [comment]
@cschlechty this is because what you refer to as the "launching control" isn't just a launching control, but literally a listbox with exactly 1 element (which you can use the up/down arrow keys to change the selected option). 

If we were to remove it, it will cause further regressions particularly around pressing up/down key without ever "launching" or opening the dropdown with an expanded list of options.

**Comment by natalieethell** [comment]
@cschlechty yeah, as Cliff mentioned, it looks like removing aria-posinset and aria-setsize when the dropdown is collapsed would cause regressions in the Narrator screen reader read out. Instead of "contains 10 items", it reads out "selection contains 1 items". When using the up and down arrows to go through the options, it also reads out "1 of 1" instead of the correct position, which might be worse than it just reading out nothing.

**Comment by natalieethell** [comment]
I'll reach out to the Edge PM I've been working with on this to see what she thinks about the "Selection contains 10 items" issue.

**Comment by natalieethell** [comment]
I spoke with the Edge PM and she makes a good point about this new behavior. 

> I actually think that the behavior of dropping posinset in this case probably makes sense too. Previously, when the event was being raised from the option, you were getting info about that option and its context within the parent control. Now that the event is being raised from the composite control, you’re getting an overview about that control: what its name is, its state (the current selection), the fact that it’s required, and how many options it contains. Then as you drill down into the options you’d get info about their context again. This seems like a good behavior to me

This especially makes sense when you think of the read-out as "Basic dropdown example, Option a selected, selection, contains 10 items,". The extra comma makes a difference.

**Comment by KevinTCoughlin** [comment]
Good discussion all 👏

File: packages/office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  public render(): JSX.Element {
+    return (
+      <div className="docs-DropdownExample">
+        <Dropdown
+          placeholder="Select an Option"
+          label="Required dropdown example:"
+          id="Requireddrop1"
+          ariaLabel="Required dropdown example"
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx]: Do we need an aria label on this if it's the exact same as the regular component label?

File: packages/office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    super(props);
+  }
+
+  public render(): JSX.Element {
+    return (
+      <div className="docs-DropdownExample">
+        <Dropdown
+          placeholder="Select an Option"
+          label="Required dropdown example:"
+          id="Requireddrop1"
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx]: Is there a reason we need explicit IDs on the examples? If not, might be better to remove it and just let the component auto-generate IDs for the label/field--I noticed a problem with the ComboBox examples last week where examples got copy-pasted (then slightly modified) but all had the same ID.

Comment (natalieethell) [reply packages/office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx]: Not sure we do, I just followed the pattern in the other example files, all of which specify the ids. I can remove this one though!

File: packages/office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx (hunk 1 above)
Comment (natalieethell) [reply packages/office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx]: Good point, we probably don't. Looks like we have this pattern in the other examples as well, just FYI.

**Pull request merged**